File: problems/T5.m
function [n,m,p,q,f,g,Df,Dg,Aineq,bineq,Aeq,beq,lb,ub,x0,is_convex,is_quadratic] = T5(~)
%T5 A tri-objective test instance
%   This example was taken from:
%   Marianna De Santis, Gabriele Eichfelder, Julia Niebling, Stefan
%   Rocktäschel. Solving Multiobjective Mixed Integer Convex Optimization
%   Problems, SIAM Journal on Optimization (2020)

% Dimension of decision and criterion space
n = 3; % Continuous variables
m = 1; % Integer variables
p = 3; % Dimension criterion space
q = 1; % Number of constraints

% Problem type
is_convex = true;
is_quadratic = true;

% Objective function
f = @(x) [x(1)+x(4);x(2)-x(4);x(3)+x(4)^2];
Df = @(x) [1,0,0,1;0,1,0,-1;0,0,1,2*x(4)];

% Linear constraints (Aineq*x <= bineq, Aeq*x = beq)
Aineq = [];
bineq = [];
Aeq = [];
beq = [];

% Lower and upper bounds (lb <= x <= ub)
lb = -2.*ones(4,1);
ub = 2.*ones(4,1);
% z = [-3;-3;-1];
% Z = [3;3;5];

% Start point x0
x0 = ceil((lb+ub)/2);

% Non-linear constraints (g(x) <= 0)
g = @(x) [x(1)^2+x(2)^2+x(3)^2-1];
Dg = @(x) [2*x(1),2*x(2),2*x(3),0];
end

File: python/problems/T5.py
"""
T5 is a tri-objective mixed-integer convex test instance with
quadratic objective and constraint functions.

It is taken from:
Marianna De Santis, Gabriele Eichfelder, Julia Niebling, Stefan
Rocktäschel. Solving Multiobjective Mixed Integer Convex Optimization
Problems, SIAM Journal on Optimization (2020)
"""

import numpy as np
from pyomo.environ import *

class structure():
    pass

def build_model(_):

    model = ConcreteModel()
    model_parameters = structure()

    # Dimension of decision and criterion space
    n = 3           # Continuous variables
    m = 1           # Integer variables
    p = 3           # Dimension criterion space
    q = 1           # Number of constraints

    model_parameters.n = n
    model_parameters.m = m
    model_parameters.p = p
    model_parameters.q = q

    # Problem type
    model_parameters.is_convex = True
    model_parameters.is_quadratic = True

    # Variables
    model.x = Var(range(n+m))
    for i in range(n):
        model.x[i].domain = Reals
        model.x[i].bounds = (-2,2)
    for j in range(n,n+m):
        model.x[j].domain = Integers
        model.x[j].bounds = (-2,2)

    # Objective functions
    model.f = ObjectiveList()
    model.f.add(expr = model.x[0] + model.x[3]).deactivate()
    model.f.add(expr = model.x[1] - model.x[3]).deactivate()
    model.f.add(expr = model.x[2] + model.x[3] ** 2).deactivate()

    # Derivatives of objective functions
    model.Df = Expression(range(p), range(n+m))
    model.Df[0, 0] = 1
    model.Df[0, 1] = 0
    model.Df[0, 2] = 0
    model.Df[0, 3] = 1

    model.Df[1, 0] = 0
    model.Df[1, 1] = 1
    model.Df[1, 2] = 0
    model.Df[1, 3] = -1

    model.Df[2, 0] = 0
    model.Df[2, 1] = 0
    model.Df[2, 2] = 1
    model.Df[2, 3] = 2 * model.x[3]

    # Constraint functions
    model.g = ConstraintList()
    model.g.add(expr = sum(model.x[i] ** 2 for i in range(n)) - 1 <= 0)

    # Derivatives of constraint functions
    model.Dg = Expression(range(q), range(n+m))
    for j in range(n):
        model.Dg[0, j] = 2 * model.x[j]
    for j in range(n, n+m):
        model.Dg[0, j] = 0

    # Solver
    model_parameters.solver = 'gurobi'
    model_parameters.patch_solver = 'ipopt'

    # Initial enclosure
    model_parameters.L = np.array([[-3,-3,-1]])
    model_parameters.U = np.array([[3,3,5]])

    return model, model_parameters